File: Python/13linked_list_find.py


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)

def linked_list_find(head, target):
  current = head
  while current is not None:
    if current.val == target:
      return True
    current = current.next
  return False

# -----------------------------------

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)

def linked_list_find(head, target):
  if head is None:
    return False
  if head.val == target:
    return True
  return linked_list_find(head.next, target)

File: Javascript/13linkedListFind.js


// ITERATIVE
// n = number of nodes
// time:  O(n)
// space: O(1)
const linkedListFind = (head, target) => {
  // todo
  let current = head;
  while (current) {
    if (current.val === target) return true;
    current = current.next;
  }
  return false;
};

// ---------------------------- //
// RECURSIVE
// n = number of nodes
// time:  O(n)
// space: O(n)
const linkedListFind = (head, target) => {
  // todo
  if (!head) return false;
  if (head.val === target) return true;
  
  return linkedListFind(head.next, target);
};
